vpr/src/place/initial_placement.cpp

```diff
@@ -236,6 +257,191 @@ static bool is_block_placed(ClusterBlockId blk_id) {
     return (!(place_ctx.block_locs[blk_id].loc.x == INVALID_X));
 }
 
+static bool is_loc_legal(t_pl_loc loc, PartitionRegion& pr, t_logical_block_type_ptr block_type) {
+    const auto& grid = g_vpr_ctx.device().grid;
+    auto& place_ctx = g_vpr_ctx.mutable_placement();
+    bool legal = false;
+
+    //Check if the location is within its constraint region
+    for (auto reg : pr.get_partition_region()) {
+        if (reg.get_region_rect().contains(vtr::Point<int>(loc.x, loc.y))) {
+            //Check if blk_id type and the location physical types match
+            if (grid[loc.x][loc.y].type->index == block_type->index) {
+                //Check if the location has enough capacity left to place blk_id
+                if (place_ctx.grid_blocks[loc.x][loc.y].usage < grid[loc.x][loc.y].type->capacity) {
+                    if (grid[loc.x][loc.y].height_offset == 0 && grid[loc.x][loc.y].width_offset == 0) {
+                        legal = true;
+                        break;
+                    }
+                }
+            }
+        }
+    }
+    return legal;
+}
+
+static bool find_centroid_neighbor(t_pl_loc centroid_loc, t_logical_block_type_ptr block_type, int& cx_to, int& cy_to) {
+    const auto& compressed_block_grid = g_vpr_ctx.placement().compressed_block_grids[block_type->index];
+
+    //Determine centroid location in the compressed space of the current block
+    int cx_centroid = grid_to_compressed_approx(compressed_block_grid.compressed_to_grid_x, centroid_loc.x);
+    int cy_centroid = grid_to_compressed_approx(compressed_block_grid.compressed_to_grid_y, centroid_loc.y);
+
+    //Determine rlim in each direction
+    int first_rlim = 15;
+    int rlim_x = std::min<int>(compressed_block_grid.compressed_to_grid_x.size(), first_rlim);
+    int rlim_y = std::min<int>(compressed_block_grid.compressed_to_grid_x.size(), first_rlim);
+
+    //Determine the valid compressed grid location ranges
+    int min_cx, max_cx, delta_cx;
+    int min_cy, max_cy;
+
+    min_cx = std::max(0, cx_centroid - rlim_x);
+    max_cx = std::min<int>(compressed_block_grid.compressed_to_grid_x.size() - 1, cx_centroid + rlim_x);
+
+    min_cy = std::max(0, cy_centroid - rlim_y);
+    max_cy = std::min<int>(compressed_block_grid.compressed_to_grid_y.size() - 1, cy_centroid + rlim_y);
+
+    delta_cx = max_cx - min_cx;
+    bool legal = false;
```

Comment:
Don't need legal = false; routine below always sets it.  Better to delete this line.

Reply:
Context on what this routine is doing, why cx_from and cy_from are -1 etc. would be helpful.